Parameterize `collect_delay`

staticfloat opened the pull request:
It is occasionally helpful to slow down the GC train, as we may be
running jobs again and again, so let the user specify a GC interval in
seconds.

staticfloat committed:
Parameterize `collect_delay`

It is occasionally helpful to slow down the GC train, as we may be
running jobs again and again, so let the user specify a GC interval in
seconds.

staticfloat commented:
I'm actually not even sure we want to default to a collect delay of 0.  If we do, and a project has multiple `julia` plugin instantiations (say, one to build docs, and one to run tests) they will clear out eachother's packages every single time they run.

I kind of think we probably want to default this to something like 1 week.  Most projects should be running a full CI round at least once a week, I'd think.  What are your thoughts?

maleadt commented:
I also remembered why I was calling `GC.gc` twice: https://github.com/JuliaCI/julia-buildkite-plugin/issues/17#issuecomment-895003469

staticfloat commented:
A `collect_delay` of zero causes immediate reaping of things, even things that are only noticed as orphaned on that GC cycle.  Here's a MWE for you:

```julia
using Pkg, Dates, Test

# Download socrates artifact; it is not referred to by any `Artifacts.toml` files,
# so it is available to be immediately culled
@info("Initial install")
hash = Base.SHA1("43563e7631a7eafae1f9f8d9d332e3de44ad7239")
url = "https://github.com/staticfloat/small_bin/raw/master/socrates.tar.gz"
Pkg.Artifacts.download_artifact(hash, url)
@show Pkg.Artifacts.artifact_exists(hash)

@info("First GC, non-zero collect_delay prevents immediate deletion")
Pkg.gc(collect_delay=Second(1), verbose=true)
@show Pkg.Artifacts.artifact_exists(hash)

sleep(1)
@info("Second GC, non-zero collect_delay elapsed, item is deleted")
Pkg.gc(collect_delay=Second(1), verbose=true)
@show Pkg.Artifacts.artifact_exists(hash)



@info("Re-install")
Pkg.Artifacts.download_artifact(hash, url)
@show Pkg.Artifacts.artifact_exists(hash)

@info("Single GC, zero collect_delay causes immediate deletion")
Pkg.gc(collect_delay=Second(0), verbose=true)
@show Pkg.Artifacts.artifact_exists(hash)
```


Running it with a temporary depot gives the following:
```
$ export JULIA_DEPOT_PATH=$(pwd)/depot; rm -rf ${JULIA_DEPOT_PATH}; julia test.jl
[ Info: Initial install
Pkg.Artifacts.artifact_exists(hash) = true
[ Info: First GC, non-zero collect_delay prevents immediate deletion
      Active manifest files: 0 found
      Active artifact files: 0 found
      Active scratchspaces: 0 found
     Deleted no artifacts, repos, packages or scratchspaces
Pkg.Artifacts.artifact_exists(hash) = true
[ Info: Second GC, non-zero collect_delay elapsed, item is deleted
      Active manifest files: 0 found
      Active artifact files: 0 found
      Active scratchspaces: 0 found
     Deleted `/tmp/tmp.hzJ8zlj0ed/depot/artifacts/43563e7631a7eafae1f9f8d9d332e3de44ad7239` (211.000 byte)
     Deleted 1 artifact installation (211.000 byte)
Pkg.Artifacts.artifact_exists(hash) = false
[ Info: Re-install
Pkg.Artifacts.artifact_exists(hash) = true
[ Info: Single GC, zero collect_delay causes immediate deletion
      Active manifest files: 0 found
      Active artifact files: 0 found
      Active scratchspaces: 0 found
     Deleted `/tmp/tmp.hzJ8zlj0ed/depot/artifacts/43563e7631a7eafae1f9f8d9d332e3de44ad7239` (211.000 byte)
     Deleted 1 artifact installation (211.000 byte)
Pkg.Artifacts.artifact_exists(hash) = false
```

maleadt commented:
Right, so don't we want to run the GC twice for non-zero collect intervals?

It's not clear to me how exactly that logic works though, across Julia versions, using the CI set-up we're using (i.e., starting with a clean depot at the start of each job), but I now notice that the CUDA.jl CI jobs re-download the lazy CUDA artifacts *almost* every time. That's obviously not ideal, since these artifacts are multiple gigabytes large...

staticfloat commented:
> Right, so don't we want to run the GC twice for non-zero collect intervals?

If your `collect_delay` is reasonably large, (like, say, 1 week) running it twice shouldn't really do anything.  The only time I think you want to run it twice is when you want to remove something immediately, in which case you should just use a `collect_delay` of 0.

> I now notice that the CUDA.jl CI jobs re-download the lazy CUDA artifacts almost every time.

If they're getting removed, this is probably because there's no `Artifacts.toml` left on-disk that refers to them at Pkg.gc() time.  You should leave one of those on-disk for the GC to see.  If you do that, then they will survive until the next CI run of that same pipeline, which will then clear them out.  If you set `collect_delay` to something like 1 week, then they'll even survive other pipeline runs for up to one week.

Thinking about this more, I really think we need to have a non-zero `collect_delay` by default, as otherwise you'll have a nice caching setup right up until you add a `docs` deployment step, and suddenly because your `docs` deployment doesn't use anything from your main pipeline, it completely clears out your depot, and your cache gets thrashed every time.

staticfloat committed:
Default `collect_delay` to 1 week

maleadt commented:
> Thinking about this more, I really think we need to have a non-zero `collect_delay` by default, as otherwise you'll have a nice caching setup right up until you add a `docs` deployment step, and suddenly because your `docs` deployment doesn't use anything from your main pipeline, it completely clears out your depot, and your cache gets thrashed every time.

Yeah, I agree. LGTM then.

I'm not sure why it still does clean them out in the CUDA.jl case though. All my jobs do instantiate the package in some way or another, so its Artifacts.toml should keep those artifacts alive. Does it matter that those artifacts are lazy, and that they aren't actually used by all jobs (e.g., some use a system CUDA installation)?

staticfloat commented:
Can you point me to the run showing this? The last hook should show when
the artifacts get removed

On Mon, Oct 4, 2021 at 23:05 Tim Besard ***@***.***> wrote:

> Thinking about this more, I really think we need to have a non-zero
> collect_delay by default, as otherwise you'll have a nice caching setup
> right up until you add a docs deployment step, and suddenly because your
> docs deployment doesn't use anything from your main pipeline, it
> completely clears out your depot, and your cache gets thrashed every time.
>
> Yeah, I agree. LGTM then.
>
> I'm not sure why it still does clean them out in the CUDA.jl case though.
> All my jobs do instantiate the package in some way or another, so its
> Artifacts.toml should keep those artifacts alive. Does it matter that those
> artifacts are lazy, and that they aren't actually used by all jobs (e.g.,
> some use a system CUDA installation)?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaCI/julia-buildkite-plugin/pull/22#issuecomment-934094399>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA762CKXEKKL25D33P3QSLUFKISBANCNFSM5FBBQ3BA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

maleadt commented:
It's the documentation job that removes them all: https://buildkite.com/julialang/cuda-dot-jl/builds/2230#79702317-01c7-48ff-af3c-06a63891d00f

```
Running Pkg.gc()
      Active manifest files: 2 found
        `/var/lib/buildkite-agent/builds/rtx4000-gpuci4-julia-csail-mit-edu/julialang/cuda-dot-jl/docs/Manifest.toml`
        `/var/lib/buildkite-agent/builds/rtx4000-gpuci4-julia-csail-mit-edu/julialang/cuda-dot-jl/Manifest.toml`
      Active artifact files: 0 found
      Active scratchspaces: 0 found
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/CoverageTools/JSYpL` (48.859 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/GPUCompiler/DksJc` (225.604 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/NNlib/P9BhZ` (332.577 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/NNlibCUDA/EENEy` (62.293 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/0ed779f9e1d7b0c9905eb5be2c4349d32fbeb050` (1.357 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/1cd09b53b957ee12ac1f9b82ca8fa7c7976bb831` (656.279 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/492b9b82838ad8001f681afa23a408c999f00935` (736.327 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/53ec7a8febc772bd4f862f45766b4d0418f5a6ce` (5.366 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/7509a8784f10f4efba9c39feb39ebeb1730a2526` (71.510 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/793934d761bbcf52eb6fa268b61281fa8480d78f` (1.383 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/9092e536c01e35eb46f24fe641f1b284bddc492f` (101.146 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/a24a0ad892cd6eb9bd4189504dfc274450ad0312` (1.383 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/a2519a1532172a6b81faf1713dec36d71382f46c` (99.592 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/abf4b5086b4eb867021118c85b2cc11a15b764a9` (577.252 KiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/ac6c131ab02095d523f905d0501faf0966d13f5d` (1.383 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/b37afdc18b754625b23a968986ffe35252f9c875` (1.585 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/c25acff15738cdd68bddf90a063849533bd12554` (101.146 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/c3ee720712fa4156a272c5cdfd1cd3c3151dddff` (1.536 GiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/cc041ec3f165d21f041899f7cc345a3673fa7f0d` (5.366 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/e6a8966d6d03d251e98440f60020c19ad61d2ab5` (3.008 MiB)
     Deleted `~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/artifacts/eaa17e7c15ad1a27356fa2e5002f64c3096588e3` (792.607 MiB)
```

The job is here: https://github.com/JuliaGPU/CUDA.jl/blob/119daf3418d38596a66bfd35728e5dde042c1e46/.buildkite/pipeline.yml#L356-L365. It does use CUDA.jl for doctests.

staticfloat commented:
Aha, I know what the issue is here!  We're not persisting the `logs` directory in the depot, so there's no orphanage file and no `*_usage.toml` files left in the depot.  For many usages this is not a problem, as the `Pkg.add()` or `download_artifact()` commands will end up writing something out, but for this job in particular, it must have had everything already installed so no `add()` needed to happen, and we ended up not writing out any usage files at all, which then caused `gc()` to think there are no active `Artifacts.toml` files.  So I think we should also default to persisting the `logs` directory.

staticfloat committed:
Update pre-command

maleadt commented:
Ah awesome, thanks for taking a look!

maleadt reviewed (approved):
LGTM, let's give it a try :slightly_smiling_face:
